### ISSUE TITLE
[FSDP2] Eager-Mode Execution Tracker

### ISSUE THREAD

**Issue by awgu** [pytorch/pytorch#120003]
## Work Items
* Meta-device initialization / `_apply()` methods
  - [ ] https://github.com/pytorch/pytorch/issues/113045 @wz337 
  - [ ] Check compatibility with `swap_tensors` path for `_apply()`
  - **Outcome:** Once we have `DTensor` manage the padded storage, then FSDP only needs to maintain a reference to the `DTensor`, not its `_local_tensor`. We should be able to do meta-device initialization via `nn.Module.to_empty()` (using `swap_tensors` path) followed by normal initialization ops (e.g. `torch.nn.init.<...>`) that runs through `DTensor`'s op dispatch.
* `clip_grad_norm_()` / sharded gradient scaler
    - [ ] Design approach for `DTensor`-based `clip_grad_norm_()`, e.g. implementing `torch.linalg.vector_norm()`
    - [ ] Design approach for `DTensor`-based sharded gradient scaler
    - **Outcome:** We can replace existing FSDP with per-parameter FSDP in torchtrain.
* 2D sharded state dict
  - [ ] Implement strided sharding placement to handle `(Shard(0), Shard(0))` placements where FSDP shards dim-0 and TP shards dim-1 @wz337 
  - [ ] Validate integration with distributed checkpointing @wz337 
  - **Outcome:** We can do 2D training including checkpointing without communication.
* `distribute_tensor()`
  - [ ] Allow `distribute_tensor()` to take in a `DTensor` to simplify/robust-ify construction of sharded parameter and sharded post-forward parameter @wz337 
  - **Outcome:** We remove all custom `DTensor` construction code from per-parameter FSDP (`_init_sharded_param`, `_init_sharded_post_forward_param_metadata`, `to_sharded_post_forward`).
* Optimizer <> `DTensor`
  - [ ] Enable `foreach=True` by default for Adam/AdamW @wz337 
  - [ ] Support `fused=True` for Adam/AdamW @wz337 
  - [ ] Support `foreach=True` for all torch-native optimizers @wz337 
  - **Outcome:** We have competitive optimizer performance on common optimizers without user-code changes (like passing `foreach=True` explicitly).
* Reduce-scatter copy-in kernel/fast-path
  - [ ] Add new aten op for pad-cat @BoyuanFeng 
  - **Outcome:** We can achieve competitive performance with existing FSDP even on low-compute-density workloads (e.g. recommendation models).
* FSDP extensions
  - [ ] Add pre/post-all-gather extensions support @awgu 
  - [ ] Add pre/post-all-gather extensions for float8_experimental @awgu 

## References
RFC: https://github.com/pytorch/pytorch/issues/114299

**Comment by ad8e** [comment]
fp16 is used for diffusion models, so the lack of gradient scaler is a blocker there.

However, https://github.com/pytorch/pytorch/pull/116054 seems relevant, which would complicate work.

**Comment by awgu** [comment]
Are there any example open source repos for training diffusion models in fp16?

We can try to get some bandwidth to look at `DTensor` + grad scaler.

**Comment by Birch-san** [comment]
edm2 trains a diffusion model in fp16 (albeit with some quirks like matmuling with weights which have been normalized in fp32 then downcasted back to fp16). it's a reasonably simple codebase:  
https://github.com/Nvlabs/edm2  
https://arxiv.org/abs/2312.02696

likewise the original OpenAI UNet (ADM) trained in fp16:  
https://github.com/openai/guided-diffusion  
https://arxiv.org/abs/2105.05233  
> We train in 16-bit precision using loss-scaling, but maintain 32-bit weights, EMA, and optimizer state

to be honest, those are actually the only examples I can think of. maybe some parameter-efficient finetunes train in fp16.

HDiT can train for a while in mixed-precision float16, but eventually the colours will fry:  
https://github.com/crowsonkb/k-diffusion  
but the toy configs (Oxford Flowers, CIFAR-10) should survive for a few minutes of training, if you're just trying to do a test and see "does it get closer to looking like an image".

**Comment by ad8e** [comment]
Update: My own fp16 grad scaler is 50 lines. It doesn't handle the fancy mixed-device things that the PyTorch grad scalar does. It's somewhat faster by avoiding unnecessary kernel launches (filed in another bug) and a CPU-GPU sync (pulling the scale from the GPU each step for logging). So if anyone wants to use fp16, it is easy to roll a good scaler from scratch.

**Comment by weifengpy** [comment]
> Update: My own fp16 grad scaler is 50 lines. It doesn't handle the fancy mixed-device things that the PyTorch grad scalar does. It's somewhat faster by avoiding unnecessary kernel launches (filed in another bug) and a CPU-GPU sync (pulling the scale from the GPU each step for logging). So if anyone wants to use fp16, it is easy to roll a good scaler from scratch.

curious: are you operating directly on DTensor._local_tensor or going through DTensor dispatch ?

**Comment by ad8e** [comment]
Currently I use `to_local()`, then manually square and all-reduce the norms. There was some issue with calling `torch._foreach_norm` on DTensor directly, but maybe it was my fault, I didn't diagnose carefully.

**Comment by weifengpy** [comment]
> Currently I use `to_local()` in a no-grad context, then `torch._foreach_norm`, then `torch.stack()`, then manually square and all-reduce the norms in one go. There was some issue with calling `torch._foreach_norm` on DTensor directly, but maybe it was my fault, I didn't diagnose carefully.
> 
> I use the per-tensor grad norms for other purposes too, so my format of having all the grad norms in a single tensor fits the other stuff I'm doing.

thanks for the details. Great work!

**Comment by ad8e** [comment]
I tested `_foreach_norm(DTensor)`; the prior error disappeared. It's slower than using local tensors however. https://api.wandb.ai/links/novelaix/va6xg9g8

Interestingly, the `train/overhead` panel also shows a slowdown - but this measures time doing Python loop things (wandb, async dataloader, non-syncing if()s), and excludes the optimizer and model. So the expectation is that changing the grad norm calculation shouldn't affect this overhead. Apparently not.
```
def square_norm_list(list_of_tensors):
    if isinstance(list_of_tensors[0], DTensor):
        list_of_tensors = [p.to_local() for p in list_of_tensors]
    local_norms = torch.stack(torch._foreach_norm(list_of_tensors))
    local_norms.pow_(2)
    dist.all_reduce(local_norms, async_op=False)
    return local_norms

...

    if self.distributed:
        # this is substantially faster than using DTensor dispatch for _foreach_norm
        all_grad_norm_squared = square_norm_list(all_grads)
        if self.return_norm or inv_scale is not None:
            all_grad_norm = all_grad_norm_squared.sqrt()
    else:
        all_grad_norm = torch.stack(torch._foreach_norm(all_grads))
```

**Comment by awgu** [comment]
@nighting0le01 you will get better help if you can post an issue in https://github.com/pytorch/ao with more details.

**Comment by qsh-zh** [comment]
@awgu  thanks for your amazing works.

I have some questions regarding FSDP2 stability and recommended practices

1. **API Stability Concerns**
   - I notice that the main implementations are located in `torch.distributed._composable.fsdp`
   - The underscore prefix in `_composable` typically indicates internal/unstable APIs
   
2. **Future Release Plans**
   - What is the timeline for a stable FSDP2 release?
   - Will FSDP2 replace FSDP1 directly, or will both APIs coexist?
   
3. **Production Readiness**
   - What is the current maturity level of FSDP2?
   - Are there any potential maintenance considerations for users adopting FSDP2 in its current state?

Context: As a potential user of FSDP2, I'm trying to evaluate the trade-offs between adopting the new features versus the stability risks and maintenance overhead.

**Comment by awgu** [comment]
@qsh-zh Thanks for your interest in FSDP2! Your concerns make a lot of sense to me.

**API Stability**
IIUC, PyTorch has a feature classification of prototype -> beta -> stable. FSDP2 is currently prototype. We plan to move it to a public folder soon (targeting in time for PyTorch 2.6 branch cut), and we can propose it to beta release.

The API itself is mostly stable since we have some internal adoption already (which requires some degree of stability). The core FSDP `reshard_after_forward=True` and `reshard_after_forward=False` plus bf16 mixed precision API should not change. Minor things that may change may be around the `fsdp_pre_all_gather`/`fsdp_post_all_gather` extension point for customizing the all-gather, but we will try to do so in a backward compatible way if we change it.

**Future Release Plans**
We will go through prototype -> beta -> stable release per normal. I think FSDP1 is still beta 😆 ; we will push for FSDP2 in beta soon, either in 2.6 or 2.7.
FSDP1 has much more adoption than FSDP2, which means more battle-testing, but we tried our best to transfer our learnings from FSDP1 to FSDP2, which may allow FSDP2 to mature faster.

My hope is for FSDP2 to replace FSDP1 because FSDP1's flat parameter, while great for performance, causes headache for both developers and users. So far, we have been focusing on pushing FSDP2 into some key internal customers, so we have not started pushing FSDP2 in open source yet. We have been able to achieve on-par or better performance on both transformer-based models and recommendation models, so we are optimistic we can replace FSDP1 eventually.

**Production Readiness**
Like briefly mentioned above, we have gotten FSDP2 adopted in some important internal production use cases. The production readiness should be solid. From a maintenance perspective, I think FSDP2 will have better support than FSDP1. FSDP1 has much higher complexity, and all of our bandwidth is focused on FSDP2 now.

My 2 cents is perhaps you can try out FSDP2. If you hit some issues, we can try to help. If they are just too unbearable, you can go back to FSDP1, while we try to figure out a better long term solution.

**Comment by qsh-zh** [comment]
@awgu 

Thank you for your explanations; they make a lot of sense.

I have another related question:
	•	Is the issue and track still active?
	•	Besides TorchTitan, are there any other good codebases using FSDP2 that you would recommend for reference purposes?

**Comment by awgu** [comment]
I updated the issue tracker this morning after seeing your comment. There are still a few things that may need a bit more validation, but the main items are all done. Let me know if there is something you want to see done.

Currently, I think it is mainly torchtitan using FSDP2. Ideally we could have some more examples, especially not just on Llama. That being said, switching from FSDP1 to FSDP2 is not meant to be too hard, and https://github.com/pytorch/torchtitan/blob/main/docs/fsdp.md should be helpful for that.

**Comment by awgu** [comment]
Yes, FSDP2 should address this. The memory usage is deterministic.

**Comment by qsh-zh** [comment]
@awgu awesome. Do you mind briefly explaining why fsdp1 has the issue and how fsdp2 resolves it?

**Comment by awgu** [comment]
I am out for a week so cannot give a detailed response, but you can look at the recordStream part of the RFC linked at the bottom of this original issue.
